Add support for boolean constraints (about nullness) in Contract annotations

carterkozak opened the pull request:
For example, the following method will behave equivalently to guava `Preconditions.checkState` with regards to null checks:
```java
@Contract("false -> fail")
static void check(boolean pass) {
  if (!pass) throw new RuntimeException();
}
```

This implementation works in a couple of layers:
- Firstly, contracts with a single boolean resultin in failure are handled by inserting a conditional throw node inline following the annotated method. This provides support for complex boolean logic input validateion.
- Secondly, simple null equal-to and not-equal-to checks are handled by the existing ContractHandler. These don't support such complex inputs, however they are able to work correctly with boolean outputs rather than exclusively `fail` cases.

I've updated the ContractHandler to use more precise language around antecedent nullness, previously null antecedents were described as nullable where cases required null. Please let me know if I've misunderstood the logic, but re-framing the values helped me to understand the logic.

Thank you for contributing to NullAway!

Please note that once you click "Create Pull Request" you will be asked to sign our [Uber Contributor License Agreement](https://cla-assistant.io/uber/NullAway) via [CLA assistant](https://cla-assistant.io/).

Before pressing the "Create Pull Request" button, please provide the following:

  - [X] A description about what and why you are contributing, even if it's trivial.

  - [X] The issue number(s) or PR number(s) in the description if you are contributing in response to those.

  - [X] If applicable, unit tests.

carterkozak commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +171,194 @@ public NullnessHint onDataflowVisitMethodInvocation(
     Preconditions.checkNotNull(analysis);
     Symbol.MethodSymbol callee = ASTHelpers.getSymbol(node.getTree());
     Preconditions.checkNotNull(callee);
-    // Check to see if this method has an @Contract annotation
-    String contractString = ContractUtils.getContractString(callee, config);
-    if (contractString != null && contractString.trim().length() > 0) {
-      // Found a contract, lets parse it.
-      String[] clauses = contractString.split(";");
```
The diff is unfortunate here, I hoisted this piece to `ContractUtils` to reuse between methods, which impacts indention/formatting.

carterkozak commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsTests.java:
```diff
@@ -42,9 +42,9 @@ public void basicContractAnnotation() {
             "import org.jetbrains.annotations.Contract;",
             "public class NullnessChecker {",
-            "  @Contract(\"_, null -> true\")",
+            "  @Contract(\"_, null -> true; _, !null -> false\")",
```
Both this contract and the example on line 47 are an example of the overzealous assumption I described in the PR description. A `null -> false` doesn't provide much insight into `!null -> [value]`, both can result in false. I added a more interesting test case below where a method in which either of two args is null returns false, which failed prior to this change.

I would have preferred to make this change in a separate PR, but it would have certainly conflicted with the work toward #667 (and is required for other new tests to work correctly). Regardless, I can split this piece into a separate PR if you prefer, or if the boolean contract piece requires a longer discussion. :-)

carterkozak replied:
I was incorrect here

carterkozak commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +171,194 @@ public NullnessHint onDataflowVisitMethodInvocation(
+    private Optional<Node> visit(BinaryOperationNode comparison) {
+      Node lhs = comparison.getLeftOperand();
+      Node rhs = comparison.getRightOperand();
+      boolean lhsIsNullLiteral = lhs.accept(IsNullLiteralNodeVisitor.INSTANCE, None.INSTANCE);
+      boolean rhsIsNullLiteral = rhs.accept(IsNullLiteralNodeVisitor.INSTANCE, None.INSTANCE);
```
We can do better here by matching expressions where one side evaluates to `Nullness.NULL` and the other `Nullness.NULLABLE`, allowing for slightly more complex inputs.

This PR has become broader than I'd prefer, so I'll bias toward investigating such improvements in a follow-up PR unless it's requested in this one.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -93,6 +104,59 @@ public void onMatchTopLevelClass(
+      String[] antecedent =
+          getAntecedent(clause, tree, analysis, state, callee, originalNode.getArguments().size());
+      // Find a single value constraint that is not already known. If more than one argument with
+      // unknown nullness affect the method's result, then ignore this clause.
```
```suggestion
      // unknown nullness affects the method's result, then ignore this clause.
```

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -93,6 +104,59 @@ public void onMatchTopLevelClass(
   }
 
+  @Override
+  public MethodInvocationNode onCFGBuildPhase1AfterVisitMethodInvocation(
```
Generally this method looks really good!

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +171,194 @@ public NullnessHint onDataflowVisitMethodInvocation(
+        if (valueConstraint.equals("_")) {
+          continue;
+        } else if (valueConstraint.equals("false") || valueConstraint.equals("true")) {
+          if (arg != null) {
```
I understand what is going on here with `false` and `true` antecedents at a high level.  Question: have you seen cases in your code base(s) where this kind of targeted handling of null checks passed to boolean parameters would be useful in terms of improving NullAway's inference / dataflow analysis?

msridhar reviewed (commented):
This is a great contribution!  But I do have a few questions

carterkozak commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +171,194 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      if (accessPath == null) {
+        continue;
+      }
+      if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NULL)) {
```
I believe you're right, there's a bug here -- I'll need to reason through it again

carterkozak replied:
My initial assessment was incorrect, there was in fact no bug :-)

There wasn't support for non-null antecedents, which I confused for a bug in existing code -- sorry for the confusion! Adding handling for non-null cases here provide the expected behavior. I've added additional test coverage with examples of the cases we're protecting against.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +174,200 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      } else if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being false, then the return
+        // being true implies arg may be null and we must mark it as such in the then update.
+        thenUpdates.set(
```
These cases where we set access paths to `NULLABLE.greatestLowerBound(...)` don't impact the final NullAway results, right?  I.e., they never cause NullAway to report additional bugs?  If that's true I'd prefer to skip these access, to avoid tracking unnecessary access paths in the store (for performance)

msridhar replied:
A test case here would be great; I still don't immediately see how to write one.

carterkozak replied:
I've added relevant test coverage for both `!null -> false` and boolean-based equivalent. It turned out the former was already covered by this condition above: https://github.com/uber/NullAway/blob/4136de7a12eb131020d65811df918859e9209339/nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java#L137-L142

Given that, I've implemented a similar check for the boolean path and removed the `greatestLowerBound` calls which are no longer necessary.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +174,200 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being true, then the return being
+        // false implies arg may be null and we must mark it as such in the else update.
+        elseUpdates.set(
```
Same comment re `NULLABLE.greatestLowerBound(...)`

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +174,200 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      } else if (consequent.equals("fail") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // Arg being non-null implies the method throws an exception, then we can mark it as
+        // null on both non-exceptional exits from the method.
+        bothUpdates.set(accessPath, Nullness.NULL);
```
This also does not impact NullAway results, correct?

carterkozak replied:
I don't believe that this currently impacts handling (in the sections of the NullAway codebase that I'm familiar with), however it would empower other contracts if the improvement described in this comment were implemented: https://github.com/uber/NullAway/pull/669#discussion_r985768003

I'd be happy to remove this case if you like, I can't justify keeping code that I cannot confirm in a test.
I think the two best options moving forward are either:
1. Remove this update
2. Retain the update, replacing null literal checks with null flow lookups in NullEqualityVisitor, and adding a test that doesn't pass otherwise.

msridhar replied:
I'd like to at least see a test case that would hypothetically be impacted.  Note that with `@Contract` support we are primarily (entirely?) interested in removing false positive reports, not removing false negatives.  We know NullAway is unsound in some cases and misses bugs, but I'm not too sure that working around that with `@Contract` annotations is a great idea, in terms of understandability of what the tool is doing.  I could be convinced otherwise.

carterkozak replied:
I've added a new test case which shows how this avoids false positives in unreachable code based on nullness with inline comments describing the setup: https://github.com/uber/NullAway/blob/724a5ab383d33482abe985a97f3d3807bb2cbf5c/nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java#L357-L382

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,351 @@
+import java.util.Arrays;
+import org.junit.Test;
+
+public class NullAwayContractsBooleanTests extends NullAwayTestsBase {
```
These tests look great!

lazaroclapp replied:
Overall, this is an _impressive_ test suite. My only worry here is that all tests are on APs that are also arguments or local variables. I wonder how well the logic works (particularly the parts that check for `Nullness.NULL` rather than `Nullness.NULLABLE`) with other tracked expressions / access-paths. Could we have just a few cases calling "assumed getters" (no arg methods) or `Map::get` with constant keys?

carterkozak replied:
Good idea, I've updated a few cases with otherwise equivalent code using `Map::get` and an assumed-getter

carterkozak commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +174,200 @@ public NullnessHint onDataflowVisitMethodInvocation(
-                        state,
-                        null));
+          }
+          Node argument = node.getArgument(i);
```
Good idea

carterkozak commented:
Thank you for another great round of review! I've pushed some javadoc updates, but haven't had time to address all comments quite yet -- I'll try to get to them either tonight or in the next couple days based on my schedule :-)

carterkozak commented:
Thanks, @msridhar, I think I'm ready for another round of review so long as you're happy with the answers I've provided to your questions (and the linked example test that I've included in the latest commit).

Given the size of this PR, would it be more convenient if I split out the code which inserts conditional throw (and relevant test coverage) into a separate PR and stacked the rest on top of it?

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,436 @@
+            "package com.uber;",
+            "class Test {",
+            "  String test(Object required) {",
+            "    return Validation.identity(required == null)",
```
Can't do a full review tonight, but quick question.  Would NullAway report an error for this case before your PR?  If so I would be surprised.

carterkozak replied:
> Can't do a full review tonight

No worries, I'm not in any rush

> Would NullAway report an error for this case before your PR? If so I would be surprised.

It would not because the contract would not be parsed, and values would be derived the method parameter which is declared non-null. However, with earlier revisions of my PR, the contract handler would inspect this invocation resulting in `thenUpdates.set(accessPath, Nullness.NULLABLE);` without taking into account that `required == null` can be proven always `false`.
Commenting out this block will effectively revert the relevant piece to reproduce the issue: https://github.com/uber/NullAway/blob/5e9f10d7f9e49b8e9b4658b4c678af8bcb7995c2/nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java#L223-L227

lazaroclapp replied:
So, this detection is useful iff there is a case for `[x]Updates.set(accessPath, Nullness.NULLABLE);`, right? Do we have such case? I can't come up with an example where we'd want to refine an existing unchanging AP *to* `Nullness.NULLABLE`, but I could well be missing something... 🤔 

(Either way, cool work and it's a good test, we definitely don't want to introduce new errors here! But, I wonder if the idea to handle all cases and converses introduced some complexity that is not needed. We've been mostly thinking of `@Contract` callsites as a way to refine `NULLABLE` into `NONNULL`. There is an argument for allowing them to refine `NULLABLE` into `NULL` - but we generally don't track must-be-null - but overall little reason to refine to `NULLABLE`: `NULLABLE`->`NULLABLE` is _redundant_ and `NONNULL`->`NULLABLE` should be _impossible_...)

carterkozak replied:
Interestingly, test cases `invertIsNull` and `invertNotNull` rely on this to report an error, otherwise it seems the conditional is assumed to act as a null check (I haven't debugged to precisely where this occurs). However, I agree that it's better to miss a potential NPE than create noise unnecessarily.

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      if (accessPath == null) {
+        continue;
+      }
+      if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NULL)) {
```
Again, I don't think we want `Nullness.NULL`, but rather `Nullness.NULLABLE`, but let me see the test cases.

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being true, then the return being
+        // false implies arg may be null and we must mark it as such in the else update.
+        elseUpdates.set(accessPath, Nullness.NULLABLE);
```
Same issue as `!null -> false`, refining an existing AP to nullable is not really useful, unless I am missing something...

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
       }
     }
     return NullnessHint.UNKNOWN;
   }
```
As a meta-comment: this method grows increasingly long and it's basically a maze of cases, some refactoring for readability might make sense, either now or as a follow up...

carterkozak replied:
Agreed, I'd tried to avoid modification where possible, but at this point I think I've slipped into the worst case middle-ground between refactoring the existing code and introducing new behavior

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+   * {@link NotEqualNode} based on {@link #equals} being {@code true} or {@code false} respectively)
+   * against a null value, {@link Optional#empty()} is returned. For example, visiting {@code e ==
+   * null} with {@code new NullEqualityVisitor(true)} would return an {@link Optional} of node
+   * {@code e}.
```
This might benefit from being more than one single paragraph (i.e. using `<p>`)

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsTests.java:
```diff
@@ -267,4 +267,161 @@ public void customContractAnnotation() {
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test1(@Nullable Object o) {",
```
Do we want two test methods here? One to check the first argument to `bothNotNull` and another the second?

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsTests.java:
```diff
@@ -267,4 +267,161 @@ public void customContractAnnotation() {
+            "import javax.annotation.Nullable;",
+            "import org.jetbrains.annotations.Contract;",
+            "public class NullnessChecker {",
+            "  @Contract(\"!null -> false\")",
```
So, I still think this is an example of why updating on `"!null -> true|false"` is not very useful. Here, either way, `required` is non-null. Is there a test case or example where we would want a different answer for a method with `@Contract(\"!null -> false\")` vs one with no contract annotation, from the part of the caller? (Separate for there being a case for validating this contract within the body of the callee, of course!)

carterkozak replied:
I agree that what it's not very useful, I added this test case in order to prove that we don't declare a `NONNULL` argument `NULLABLE` and incorrectly describe impossible dereferences.

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,436 @@
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    Validation.checkTrue(o1 != null && o1.toString().equals(\"\"));",
```
This is pretty damn cool! Do we have the check for the case where we have this same code but `||` instead of `&&`? Are we correctly detecting that `o1` may still be nullable after?

carterkozak replied:
Good idea, I've added additional cases to this test

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,436 @@
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1, @Nullable Object o2) {",
+            "    Validation.checkFalse(o1 == null || o2 == null);",
```
Similar question as above, what happens with `Validation.checkFalse(o1 == null && o2 == null)`?

carterkozak replied:
Good idea, I've added additional test coverage here as well

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,436 @@
+            "    if (checkAndReturn(o1 != null) && o2 != null) {",
+            "      return o1.toString() + o2.toString();",
+            "    } else {",
+            "      return o1.toString() + ",
```
Neat, it still remembers `o1` can't be `null`.

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,436 @@
+  }
+
+  @Test
+  public void contractUnreachablePathAfterFailure() {
```
What is this test testing that `identityIsNull()` does not? Is this meant to be:

```
            "    Validation.checkFalse(o == null);",
            "    return Validation.identity(o == null)",
            "      // No error because the then-branch is unreachable",
            "      ? o.toString()",
            "      : o.toString();",
```

(Whether or not that is actually handled or a known FP)?

carterkozak replied:
I should have done a better job adding comments inline with these tests. This is meant to ensure the `checkTrue(o == null)` refines `o` to `Nullness.NULL` and correctly flags that the is-null case will NPE on `o.toString()` while nothing is erroneously reported in the unreachable else clause.

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,436 @@
+            "    checkTrue(o1 == null, required);",
+            // o1 is guaranteed to be null based the previous check, allowing
+            // us to understand this expression as 'null == o2'.
+            "    return Validation.identity(o1 == o2)",
```
Ok, in order for this to work, we need explicit tracking of `o1` as `Nullness.NULL` not `Nullness.NULLABLE`, and we seem to have it here, but I am not sure how deep this goes... a bit worried about how well we handle this in general. What if these are more complex APs than simple variables? (map gets, getter calls, etc)

lazaroclapp reviewed (commented):
Had a chance to do a pass. Love the basic idea and some of the test cases for contracts with boolean antecedents are super impressive, IMO!

That said, overall, my main concerns are:

- Is there a case for `@Contract("!null -> true|false")` refining nullness towards `NULLABLE` (or even towards `NULL`) on the callee side? That represents a big part of the complexity in this PR, I feel, and I am not sure there is a use case (there isn't a clear one in the tests... or maybe I missed it 🙂 )
- There is change testing for `NULL` rather than for `NULLABLE`, but NullAway isn't great at tracking must-be-null info. It mostly uses `NULLABLE` or `NONNULL`. I don't see anything obvious breaking because of that change, but it makes me somewhat nervous for that reason...

ps. Also, this is the kind of change that I would feel a lot more comfortable landing with some internal testing on our codebase (in addition to NullAway's unit tests).

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      } else if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being false, then the return
+        // being true implies arg may be null and we must mark it as such in the then update.
+        thenUpdates.set(accessPath, Nullness.NULLABLE);
```
> There are two options when visiting the statement `foo(o)`, either a) `o` is nullable, or b) `o` is non-null (`o` being `Nullness.NULL` is a case of being nullable, but also exceedingly rare in our tracking). If `o` was non-null before `foo(o)`, it will remain so after. If it was nullable, then there is no point to update it to `Nullness.NULLABLE`.
> 
> So, in which case would we need this change to `thenUpdates`?
> 
> (Generally, it makes sense to mark something that was nullable previous to a method call as now known to be non-null, but it makes no sense to mark an _existing_ AP as being nullable unless the value at that AP was somehow updated, I think...)

This is exactly the issue I've been thinking about.  I think the only time an AP can go from being non-null to nullable is if it gets mutated by the callee method.  This is impossible for locals, but possible for field accesses / method call results.  NullAway unsoundly assumes invoked methods are side-effect-free and deterministic, and I think it would be very surprising to users if `@Contract` annotations were capturing side effects and leading to error messages that wouldn't be there otherwise.  In general, I think we should aim for our `@Contract` handling to strictly remove false positives, not to introduce new true positives.

msridhar commented:
@carterkozak I hope you aren't getting overwhelmed with all the feedback :-)  Taking a step back, I think it might be useful to do a version of this PR that only refines access paths to `NONNULL`, never to `NULLABLE` or `NULL`.  To me, that functionality has clear utility, and we'd still keep the new logic for handling boolean antecedents.  We could maybe then have a follow-up discussion about whether refining to `NULLABLE` or `NULL` is useful / worthwhile, and if we decide it is, add that feature in a follow-up.

Again, we really appreciate the contribution!

carterkozak commented:
Not at all, I appreciate all the feedback! I should have a chance to go through and address comments either tomorrow or Sunday.

Splitting out the least controversial pieces sounds great, I should be able to get to that around the same time.

carterkozak commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -93,6 +103,62 @@ public void onMatchTopLevelClass(
+        String valueConstraint = antecedent[i].trim();
+        if ("false".equals(valueConstraint) || "true".equals(valueConstraint)) {
+          if (arg != null) {
+            supported = false;
```
Good idea, I've added a bit of text specifying that this applies to the node-insertion method, as things are a bit fuzzier in the existing codepath (not changed by this PR) in which `!null` constraints are equivalent to wildcard `_` in cases where the parameter is known non-null.

carterkozak commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+          // isNullTarget is equivalent to 'null ->' while notNullTarget is equivalent
+          // to '!null ->'. However, the valueConstraint may reverse the check.
+          boolean inverted = isNullTarget.isPresent() == valueConstraint.equals("false");
+          Nullness antecedentNullness = inverted ? Nullness.NONNULL : Nullness.NULL;
```
Writing the steps out to simplify from the inputs, to expectations (sorry if the steps seems simple, just trying to make sure I understand the question :-)

`notNullTarget.present() == true` means we have an argument of the form `obj != null`
`valueConstraint.equals("true")` means the contract includes `true -> [result]`
`inverted` is set as the result of `isNullTarget.isPresent() == valueConstraint.equals("false")` which evaluates to
```java
// isNullTarget.isPresent() must be false when notNullTarget.present()
false == valueConstraint.equals("false")

// valueConstraint.equals("true") means valueConstraint.equals("false") must be false
false == false

// Which evaluates to
true

// since inverted is true, the following evaluates to `NONNULL`
Nullness antecedentNullness = inverted ? Nullness.NONNULL : Nullness.NULL;
```

I think it's a bit more confusingly written after I swapped the old implementation out due to `Optional.isEmpty` not existing in jdk8.

I've updated this section with a truth table because boolean logic is difficult to reason about.

msridhar replied:
I agree the more explicit logic is more readable and that there shouldn't be a significant performance cost

carterkozak replied:
Good idea, I've updated the implementation with more explicit logic. I tried extracting the logic to a standalone function, but found the indirection made things a bit harder to follow (but I defer to your judgement if you'd prefer a separate function).

carterkozak commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
           }
-          continue;
+          arg = node.getArgument(i);
+          argAntecedentNullness = valueConstraint.equals("null") ? Nullness.NULL : Nullness.NONNULL;
```
I think this comment also helps me understand your comment in https://github.com/uber/NullAway/pull/669#discussion_r989495957

I believe there are two things going on: First, we're not confident about trusting the value of `Nullness.NULL` as an input to checks, which I'll pull out to another issue or PR for discussion separately. Second, the reliance on `Nullness.NULL` with regard to the `argAntecedentNullness` variable.

I think we're in agreement about trimming out the `NULL` pieces as inputs to checks in this pull request, however I don't believe `argAntecedentNullness` `Nullness` values themselves are connected to dataflow analysis at all beyond reusing the same data-type -- this was confusing for me when I began investigating this source file because the argAntecedentNullness represents a constraint, which is either `NULL` or `NONNULL`, where we can't reason about cases in which the arg is neither, if that makes sense. To summarize, we could replace all the assignments and checks of `Nullness.NULL` specifically on `argAntecedentNullness` with `Nullness.NULLABLE` (or even Nullness.BOTTOM, for that matter), and the results would not be impacted because the value of that variable is confined to this method.

carterkozak commented:
I haven't had a chance to address all the comments yet, I'll try to finish them out in the next couple days.
My plan is to iterate forward on this branch to ensure I fully understand your comments, then split out the non-controversial piece as discussed, retaining as much of the test coverage as possible.

carterkozak commented:
I think I've responded to the comments, please let me know if I missed anything!

I've updated this PR to only refine into `NONNULL`, this only impacted three test cases: 63ec9dbe29df899d7eab81b4961eabe6378991eb
The boolean-comparison test is still using `inputs.valueOfSubNode(lhs|rhs)` and comparing that against `Nullability.NULL` as it's simpler than using another visitor, and works in the cases that I'm interested in (null literal), but I can bring back the visitor which searched for a NullLiteralNode if you prefer.

Thanks again for such thorough and well-reasoned reviews!

msridhar commented:
FYI I will get to taking another look here.  It's just a bit more complex contribution so I need to block out a bit of time :-)

schlosna commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,408 @@
+            "  }",
+            "  @Contract(\"true -> fail\")",
+            "  static void checkFalse(boolean value) {",
+            "    if (!value) throw new RuntimeException();",
```
I think this should throw when `value == true`

```suggestion
            "    if (value) throw new RuntimeException();",
```

carterkozak replied:
You're absolutely right, thanks!

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -93,6 +103,70 @@ public void onMatchTopLevelClass(
+        if ("false".equals(valueConstraint) || "true".equals(valueConstraint)) {
+          if (arg != null) {
+            // We don't currently support contracts depending on the boolean value of more than one
+            // argument
+            // using the node-insertion method.
```
fix comment formatting

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +181,203 @@ public NullnessHint onDataflowVisitMethodInvocation(
+          getAntecedent(clause, tree, analysis, state, callee, node.getArguments().size());
+      String consequent = getConsequent(clause, tree, analysis, state, callee);
+
+      // Find a single value constraint that is not already known. If more than one arguments with
```
```suggestion
      // Find a single value constraint that is not already known. If more than one argument with
```

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java:
```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+          // to '!null ->'. However, the valueConstraint may reverse the check.
+          boolean inverted = isNullTarget.isPresent() == valueConstraint.equals("false");
+          Nullness antecedentNullness = inverted ? Nullness.NONNULL : Nullness.NULL;
+          Nullness targetNullness = inputs.valueOfSubNode(nullTestTarget);
```
@carterkozak is this concern no longer present given the reduced scope of this PR?

carterkozak replied:
I think so, if I understand the concern, but the system is complex and we are in a way relying on the value here in a way: not to produce a static error, but rather to whittle down statically-known constraints until there's only one left that's not guaranteed true.

I could update the following line to `antecedentNullness.equals(targetNullness) && !antecedentNullness.equals(Nullness.NULL)` to prevent us from taking `x == null` as true where `final Object x = null` into account, while still matching behavior of the guaranteed-`!null` contract. What do you think?

msridhar replied:
> I think so, if I understand the concern, but the system is complex and we are in a way relying on the value here in a way: not to produce a static error, but rather to whittle down statically-known constraints until there's only one left that's not guaranteed true.

Can you point me at the test case that takes advantage of this reasoning?  I think I understand, just want to make sure.

msridhar replied:
Nevermind, I think it's the new test you added?  `nonNullBooleanDoubleContractTest()`?

carterkozak replied:
That's correct -- we're able to rule out the second parameter because it's guaranteed to satisfy the contract based on the object being non-null.

msridhar replied:
Ok, so I think I get it.  The idea is that if we have an expression `e` of `@NonNull` type, and there is a check `e == null`, we know for sure that the check will evaluate to false.  So, if that check expression is the first argument to a function with contract `false, false -> fail`, we can prove that the failure case will occur iff the second argument is false.  So, continuing the example, if the second argument is `e2 == null` and `e2` is `@Nullable`, we can refine `e2` to `@NonNull` in the non-failure case.  Does that sound right?

Given this reasoning is localized to this function, I'm ok with keeping it as-is.  @lazaroclapp can give his opinion when he returns.

carterkozak replied:
That is exactly right

carterkozak replied:
I don't believe we track `Nullness.NULL` from fields, only `NULLABLE` -- I've added test coverage for local variables, which track explicit null assignment, but not fields as they wouldn't pass (and I don't think we want to verify that fields cannot produce `NULL` in the scope of this feature).

msridhar reviewed (commented):
Overall this looks really good to me now; I only have a couple of very minor comments.  We will wait until @lazaroclapp returns and can complete his review and test this internally before we land it

msridhar reviewed (approved):
This PR is approved from my perspective.  Will wait for an approval from @lazaroclapp before landing

lazaroclapp reviewed (approved):
Apologies for the delay here! Just had a chance to go over this and also run a test of NullAway with these changes on our internal codebase. 

A few nits and comments below, with a couple places where a clarifying comment might be good, but overall this looks good to land!

Also, one small request before we squash+land this: could the title be something we can directly use as a line in the changelog (following convention with our other PRs/commits). E.g. "Add support for boolean constraints (about nullness) in Contract annotations"?

msridhar commented:
Let's also merge latest master (or rebase) and make sure everything is green before landing

carterkozak commented on nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java:
```diff
@@ -0,0 +1,551 @@
+            "      ? o.toString()",
+            // This path is unreachable because o is guaranteed to be null
+            // after checkTrue(o == null). No failures should be reported.
+            "      : o.toString();",
```
Documenting test cases is always difficult, describing what the test represents vs the specifics that we want to protect against. This test involves unreachable code, however the aspect of the implementation I'm testing is that we don't produce a compile-time error in a case that cannot NPE.

carterkozak commented:
> Apologies for the delay here!

No need to apologize, thank you for another great round of review :-)

Updated based on feedback, rebased to the tip of master, and updated the PR title. Please let me know if I've missed anything!

lazaroclapp reviewed (approved):
Looks good to land to me! 🎉  @msridhar?

lazaroclapp commented:
Checked with @msridhar and landed this PR! :) 

Once again, thanks for bearing with me on the review delays and multiple rounds of review and thanks for the cool feature @carterkozak !

msridhar commented:
Thanks again for the awesome contribution @carterkozak!

msridhar commented:
@carterkozak would it be helpful to you to have a release soon that incorporates this change?

carterkozak commented:
My pleasure, thank you for maintaining NullAway! A release would be fantastic if it's not too much trouble, both for this feature and the `CheckOptionalEmptiness` NPE fix

lazaroclapp commented:
Just FYI, cutting a release right now containing this PR and #682. It should propagate to Maven Central within the next few hours!

carterkozak commented:
Thanks!
